### PR TITLE
fix: only accept dates that are ISO format or timestamps [DHIS2-10717]

### PR DESCRIPTION
**Pull request by jbee**
Makes the JSON => `Date` more strict so that it rejects values that are not either a valid ISO format string (as used in DHIS2) or a timestamp.

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java [hunk 1]
```diff
@@ -46,38 +50,44 @@
...
     @Test
     public void testIsoDateSupport()
-        throws JsonProcessingException
     {
-        Map<String, Date> yearTest = jsonMapper.readValue( createDateTest( "2019" ), new DateMapTypeReference() );
-        assertEquals( yearTest.get( "date" ), DateUtils.parseDate( "2019" ) );
-        Map<String, Date> yearMonthTest = jsonMapper.readValue( createDateTest( "2019-01" ),
-            new DateMapTypeReference() );
-        assertEquals( yearMonthTest.get( "date" ), DateUtils.parseDate( "2019-01" ) );
+        assertParsedAsDate( DateUtils.parseDate( "2019" ), "2019" );
```
Comment (enricocolasante) [inline dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java:58]: Maybe we can add a few test cases like:
- Complete date
- Date with time
- Date with time and timezone
- Dates in a format that is not ISO

**Review by enricocolasante — APPROVED** [review]
Just a small suggestion on the tests

File: dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java (hunk 1 above)
Comment (jbee) [reply dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java:58]: Yes, there is an open question as well: we do not accept ISO with 4 digits for nanoseconds part but that is the standard output of ISO DATE TIME format. Feels a little inconsistent.

**Review by netroms — APPROVED** [review]

File: dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java (hunk 1 above)
Comment (jbee) [reply dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java:58]: I added more test scenarios.

File: dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java [hunk 2]
```diff
@@ -46,38 +50,54 @@
...
 
     @Test
-    public void testUnixEpochTimestamp()
-        throws JsonProcessingException
+    public void testIsoDateSupport_UnsupportedFormats()
     {
-        Map<String, Date> unixEpochDateString = jsonMapper.readValue( createDateTest( "1575118800000" ),
-            new DateMapTypeReference() );
-        assertEquals( unixEpochDateString.get( "date" ), new Date( 1575118800000L ) );
+        assertNotParsedAsDate( "2019-01-01T11:55:01.4444" );
```
Comment (jbee) [inline dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java:69]: Note that DHIS2 ISO format does not allow for this which is unfortunate as this is the default output of `DateTimeFormatter.ISO_LOCAL_DATE_TIME.format`

Comment (jbee) [reply dhis-2/dhis-support/dhis-support-commons/src/test/java/org/hisp/dhis/commons/config/jackson/JacksonObjectMapperConfigTest.java:69]: Created #7726 for it.

**Pull request merged**